csrc/quantization/qqq/qqq_gemm_kernel.cu

```diff
@@ -0,0 +1,1317 @@
+  // size is not fixed, we can not shuffle before inference we shuffle it when
+  // fetching s1 from global memory to shared memory, that's why s1_sh_wr is
+  // like this
+  int s1_sh_wr =
+      (threadIdx.x / 16) * 16 + (threadIdx.x % 8) * 2 + (threadIdx.x % 16) / 8;
+  int s1_sh_rd = (threadIdx.x % 32) / 4;
+  bool s1_sh_wr_pred = threadIdx.x < prob_m;
+
+  int s2_gl_rd = s2_sh_stride * slice_col + threadIdx.x;
+  int s2_sh_wr = threadIdx.x;
+  int s2_sh_rd = 16 * ((threadIdx.x / 32) % (thread_n_blocks / 4)) +
+                 2 * ((threadIdx.x % 32) % 4);
+  bool s2_sh_wr_pred = threadIdx.x < s2_sh_stride;
+
+  int s3_gl_rd, s3_sh_wr, s3_sh_rd;
+  bool s3_sh_wr_pred;
+  if constexpr (group_blocks != -1) {
+    s3_gl_rd = s3_gl_stride * ((thread_k_blocks * slice_row) / group_blocks) +
+               s3_sh_stride * slice_col + threadIdx.x;
+    s3_sh_wr = threadIdx.x;
+    // NOTE(HandH1998): s3_sh_rd is related to mma output C
+    s3_sh_rd = 8 * ((threadIdx.x / 32) % (thread_n_blocks / 4)) +
+               (threadIdx.x % 32) / 4;
+    s3_sh_wr_pred = threadIdx.x < s3_sh_stride;
+  }
+
+  // Precompute which thread should not read memory in which iterations; this is
+  // needed if there are more threads than required for a certain tilesize or
+  // when the batchsize is not a multiple of 16.
+  bool a_sh_wr_pred[a_sh_wr_iters];
+  #pragma unroll
+  for (int i = 0; i < a_sh_wr_iters; i++)
+    a_sh_wr_pred[i] = a_sh_wr_delta * i + a_sh_wr < a_sh_stride * prob_m;
+
+  // To ensure that writing and reading A tiles to/from shared memory, the
+  // latter in fragment format, is fully bank conflict free, we need to use a
+  // rather fancy XOR-based layout. The key here is that neither reads nor
+  // writes of the 16-byte `int4` blocks of 8 consecutive threads involve the
+  // same shared memory banks. Further, it seems (based on NSight-Compute) that
+  // each warp must also write a consecutive memory segment?
+  auto transform_a = [&](int i) {
+    int row = i / a_gl_rd_delta_o;
+    return a_gl_rd_delta_o * row + (i % a_gl_rd_delta_o) ^ row;
+  };
+  // Since the computation of this remapping is non-trivial and, due to our main
+  // loop unrolls, all shared memory accesses are static, we simply precompute
+  // both transformed reads and writes.
+  int a_sh_wr_trans[a_sh_wr_iters];
+  #pragma unroll
+  for (int i = 0; i < a_sh_wr_iters; i++)
+    a_sh_wr_trans[i] = transform_a(a_sh_wr_delta * i + a_sh_wr);
+  int a_sh_rd_trans[b_sh_wr_iters][thread_m_blocks];
+  #pragma unroll
+  for (int i = 0; i < b_sh_wr_iters; i++) {
+  #pragma unroll
+    for (int j = 0; j < thread_m_blocks; j++)
+      a_sh_rd_trans[i][j] =
+          transform_a(a_sh_rd_delta_o * i + a_sh_rd_delta_i * j + a_sh_rd);
+  }
+
+  // Since B-accesses have non-constant stride they have to be computed at
+  // runtime; we break dependicies between subsequent accesses with a tile by
+  // maintining multiple pointers (we have enough registers), a tiny
+  // optimization.
```

Comment:
nit: dependicies -> dependencies and maintining -> maintaining

I think codespell should catch this, so make sure you can run `format.sh` on this to get the build green :)